### PR TITLE
add optional additional subnets to Nutanix provider

### PR DESCRIPTION
**Pull request by lucakuendig**
**What this PR does / why we need it**:
This PR adds the possibility to create a VM with an additional NIC for a storage subnet for the Nutanix provider. We need this as we use a dedicated network for our Nutanix storage provider.

**What type of PR is this?**
/kind feature

**Special notes for your reviewer**:
I did the implementation as minimal as possible. But we where diskussing different options and I am open to change the code if you require it.
Second option would have been an "additionalSubnetNamess" List instead of the storageSubnetName field. But we were not sure if somebody needs more than two NICs anyways.
Third option would be replacing subnetName with a List. But this would break backwards compatibility. 
Let me know if you think any of the two options above would be a better solution for the project.

**Does this PR introduce a user-facing change? Then add your Release Note here**:

```release-note
Add new field, additionalSubnetNames, to the Nutanix provider which allows to configure additional network interfaces. 
```

**Documentation**:

```documentation
TBD
```

### CODE REVIEW

**Comment by lucakuendig** [comment]
@ahmedwaleedmalik Thx for advancing with the PR. Just a quick question, as I understand I need somebody to review the PR to progress in the process. Are you assigning somebody or do I need to @ some people?

**Comment by embik** [comment]
Thank you for the contribution @lucakuendig!

> Second option would have been an "additionalSubnetNamess" List instead of the storageSubnetName field. But we were not sure if somebody needs more than two NICs anyways.

This would make the most sense in my opinion. I don't see anything in the code about this subnet being treated differently because it's a storage subnet, so IMHO the field name would be a bit confusing. `additionalSubnetNames` as a list seems to be the better solution. Maybe no one needs more than two subnets, but making this more flexible allows for covering more use cases.

File: test/e2e/provisioning/testdata/machinedeployment-nutanix.yaml [hunk 1]
```diff
@@ -33,6 +33,8 @@ spec:
             clusterName: '<< NUTANIX_CLUSTER >>'
             projectName: '<< NUTANIX_PROJECT >>'
             subnetName: '<< NUTANIX_SUBNET >>'
+            additionalSubnetNames:
+              - '<< NUTANIX_ADDITIONALSUBNETS >>'
```
Comment (embik) [inline test/e2e/provisioning/testdata/machinedeployment-nutanix.yaml:37]: Our e2e environment currently does not have multiple subnets, so I'm afraid we won't be able to test for this anyway. I would say we remove the references from `test/e2e/provisioning` for now and will revisit this on our end once we have the ability to integrate it into the test environment. You can however provider `additionalSubnetNames: []` in this file if you want, to. make sure we test with the field present.

**Review by embik — CHANGES_REQUESTED** [review]
Couple of review comments, but this looks great so far. Thank you for the quick change in how this PR is working!

File: pkg/cloudprovider/provider/nutanix/types/types.go [hunk 2]
```diff
@@ -38,10 +38,11 @@ type RawConfig struct {
...
 	ProxyURL      providerconfigtypes.ConfigVarString `json:"proxyURL,omitempty"`
 
-	ClusterName providerconfigtypes.ConfigVarString  `json:"clusterName"`
-	ProjectName *providerconfigtypes.ConfigVarString `json:"projectName,omitempty"`
-	SubnetName  providerconfigtypes.ConfigVarString  `json:"subnetName"`
-	ImageName   providerconfigtypes.ConfigVarString  `json:"imageName"`
+	ClusterName           providerconfigtypes.ConfigVarString   `json:"clusterName"`
+	ProjectName           *providerconfigtypes.ConfigVarString  `json:"projectName,omitempty"`
+	SubnetName            providerconfigtypes.ConfigVarString   `json:"subnetName"`
+	AdditionalSubnetNames []providerconfigtypes.ConfigVarString `json:"additionalSubnetNames,omitempty"`
```
[the remark this answers is not in the document]
Comment (lucakuendig) [reply pkg/cloudprovider/provider/nutanix/types/types.go]: Changed to []string as requested

File: pkg/cloudprovider/provider/nutanix/client.go [hunk 3]
```diff
@@ -104,6 +104,31 @@ func createVM(ctx context.Context, client *ClientSet, name string, conf Config,
...
+	nicList := []*nutanixv3.VMNic{
+		{
+			SubnetReference: &nutanixv3.Reference{
+				Kind: pointer.String(nutanixtypes.SubnetKind),
+				UUID: subnet.Metadata.UUID,
+			},
+		},
+	}
+
+	if len(conf.AdditionalSubnetNames) != 0 {
```
[the remark this answers is not in the document]
Comment (lucakuendig) [reply pkg/cloudprovider/provider/nutanix/client.go]: makes sense, did not think about hat. Removed it.

File: pkg/cloudprovider/provider/nutanix/provider.go [hunk 4]
```diff
@@ -181,6 +182,16 @@ func (p *provider) getConfig(provSpec clusterv1alpha1.ProviderSpec) (*Config, *p
 		return nil, nil, nil, err
 	}
 
+	if len(rawConfig.AdditionalSubnetNames) != 0 {
```
[the remark this answers is not in the document]
Comment (lucakuendig) [reply pkg/cloudprovider/provider/nutanix/provider.go]: Removed

File: pkg/cloudprovider/provider/nutanix/provider.go [hunk 5]
```diff
@@ -227,6 +238,14 @@ func (p *provider) Validate(ctx context.Context, spec clusterv1alpha1.MachineSpe
 		return fmt.Errorf("failed to get subnet: %w", err)
 	}
 
+	if len(config.AdditionalSubnetNames) != 0 {
```
[the remark this answers is not in the document]
Comment (lucakuendig) [reply pkg/cloudprovider/provider/nutanix/provider.go]: Removed

File: test/e2e/provisioning/testdata/machinedeployment-nutanix.yaml (hunk 1 above)
Comment (lucakuendig) [reply test/e2e/provisioning/testdata/machinedeployment-nutanix.yaml]: changed to additionalSubnetNames: [] as suggested

**Comment by lucakuendig** [comment]
> Couple of review comments, but this looks great so far. Thank you for the quick change in how this PR is working!

thx for the quick review. I made the requested changes and hope it looks good now?

**Review by embik — APPROVED** [review]
Thank you very much for your contribution @lucakuendig! the last e2e test should succeed soon, we'll keep an eye on it. It's not related to the PR though.

/approve

**Comment by lucakuendig** [comment]
hey @ahmedwaleedmalik could you have a look aswell and give your approve if you think it is ok?

**Review by ahmedwaleedmalik — APPROVED** [review]
/approve ⚡

**Pull request merged**